lib/cleanse.rb

```diff
@@ -17,108 +17,94 @@
 #
 
 module Cleanse
-  def self.cleanse(vrp, result)
-    if result
-      cleanse_empties_fills(vrp, result)
-      # cleanse_empty_routes(result)
-    end
+  def self.cleanse(vrp, solution)
+    return unless solution
+
+    cleanse_empties_fills(vrp, solution)
+    # cleanse_empty_routes(result)
   end
 
-  def self.same_position(vrp, previous, current)
-    previous.matrix_index && current.matrix_index && (vrp.matrices.first[:time].nil? || vrp.matrices.first[:time] && vrp.matrices.first[:time][previous.matrix_index][current.matrix_index] == 0) &&
-      (vrp.matrices.first[:distance].nil? || vrp.matrices.first[:distance] && vrp.matrices.first[:distance][previous.matrix_index][current.matrix_index] == 0) ||
-      previous.location && current.location && previous.location.lat == current.location.lat && previous.location.lon == current.location.lon
+  def self.same_position(vrp, current)
+    current.info.travel_time&.zero? || current.info.travel_distance&.zero?
   end
 
   def self.same_empty_units(capacities, previous, current)
     if previous && current
       if previous
-        previous_empty_units = previous.quantities.collect{ |quantity|
-          quantity.unit.id if quantity.empty
+        previous_empty_units = previous.loads.map{ |l|
+          l.quantity.unit.id if l.quantity.empty
         }.compact
       end
       if current
-        useful_units = (current.quantities.collect{ |quantity|
-          quantity.unit.id
+        useful_units = (current.loads.map{ |l|
+          l.quantity.unit.id
         }.compact & capacities)
 
-        current_empty_units = current.quantities.collect{ |quantity|
-          quantity.unit.id if quantity.empty
+        current_empty_units = current.loads.map{ |l|
+          l.quantity.unit.id if l.quantity.empty
         }.compact
       end
-      !previous_empty_units.empty? && !current_empty_units.empty? && (useful_units & previous_empty_units & current_empty_units == useful_units)
+      !previous_empty_units.empty? && !current_empty_units.empty? &&
+        (useful_units & previous_empty_units & current_empty_units == useful_units)
     end
   end
 
   def self.same_fill_units(capacities, previous, current)
     if previous && current
       if previous
-        previous_fill_units = previous.quantities.collect{ |quantity|
-          quantity.unit.id if quantity.fill
+        previous_fill_units = previous.loads.map{ |l|
+          l.quantity.unit.id if l.quantity.fill
         }.compact
       end
       if current
-        useful_units = (current.quantities.collect{ |quantity|
-          quantity.unit.id
+        useful_units = (current.loads.map{ |l|
+          l.quantity.unit.id
         }.compact & capacities)
 
-        current_fill_units = current.quantities.collect{ |quantity|
-          quantity.unit.id if quantity.fill
+        current_fill_units = current.loads.map{ |l|
+          l.quantity.unit.id if l.quantity.fill
         }.compact
```

Comment:
Can you replace `l` with `load`, it is harder to read since it looks like :open_mouth:  `1.quantity`

Reply:
as load is a method of ruby Kernel, i will avoid to use it, may `loa` suits ?

Reply:
ah :face_exhaling:  effectively.. in that case, as you wish, you can ignore my comment.